app/frontend/wallet/shelley/shelley-ledger-crypto-provider.ts

```diff
@@ -220,80 +219,164 @@ const ShelleyLedgerCryptoProvider = async ({
+        path,
+      },
     }
   }
 
-  function prepareStakepoolRegistrationCertificate(
-    certificate: TxStakepoolRegistrationCert,
+  function prepareRelays(relays: TxStakepoolRelay[]): LedgerTypes.Relay[] {
+    return relays.map((relay) => {
+      switch (relay.type) {
+        case TxRelayType.SINGLE_HOST_IP:
+          return {
+            type: LedgerTypes.RelayType.SINGLE_HOST_IP_ADDR,
+            params: relay.params,
+          }
+        case TxRelayType.SINGLE_HOST_NAME:
+          return {
+            type: LedgerTypes.RelayType.SINGLE_HOST_HOSTNAME,
+            params: relay.params,
+          }
+        case TxRelayType.MULTI_HOST_NAME:
+          return {
+            type: LedgerTypes.RelayType.MULTI_HOST,
+            params: relay.params,
+          }
+        default:
+          throw NamedError('InvalidRelayType')
+      }
+    })
+  }
+
+  function preparePoolOwners(
+    stakepoolOwners: TxStakepoolOwner[],
+    stakingAddress: Address,
     path: BIP32Path
-  ): LedgerCertificate {
-    const {data} = bech32.decode(certificate.stakingAddress)
-    const poolOwners = certificate.poolRegistrationParams.poolOwners.map((owner) => {
-      return !Buffer.compare(Buffer.from(owner.stakingKeyHashHex, 'hex'), data.slice(1))
-        ? {stakingPath: path}
-        : {...owner}
+  ): LedgerTypes.PoolOwner[] {
+    const {data: stakingAddressBuff} = bech32.decode(stakingAddress)
+    const poolOwners: LedgerTypes.PoolOwner[] = stakepoolOwners.map((owner) => {
+      // TODO: helper function for slicing first bit from staking address so its stakingKeyHash
+      return !Buffer.compare(
+        Buffer.from(owner.stakingKeyHashHex, 'hex'),
+        stakingAddressBuff.slice(1)
+      )
+        ? {
+          type: LedgerTypes.PoolOwnerType.DEVICE_OWNED,
+          params: {stakingPath: path},
+        }
+        : {
+          type: LedgerTypes.PoolOwnerType.THIRD_PARTY,
+          params: {stakingKeyHashHex: owner.stakingKeyHashHex},
+        }
     })
-    if (!poolOwners.some((owner) => owner.stakingPath)) {
+    if (!poolOwners.some((owner) => owner.type === LedgerTypes.PoolOwnerType.DEVICE_OWNED)) {
       throw NamedError('MissingOwner', {
         message: 'This HW device is not an owner of the pool stated in registration certificate.',
       })
     }
+    return poolOwners
+  }
+
+  function prepareStakepoolRegistrationCertificate(
+    certificate: TxStakepoolRegistrationCert,
+    path: BIP32Path
+  ): LedgerTypes.Certificate {
+    const {stakingAddress, poolRegistrationParams} = certificate
+    const poolOwners = preparePoolOwners(poolRegistrationParams.poolOwners, stakingAddress, path)
+    const margin = {
+      numerator: poolRegistrationParams.margin.numeratorStr,
+      denominator: poolRegistrationParams.margin.denominatorStr,
+    }
+    const metadata = poolRegistrationParams.metadata
+      ? {
+        metadataUrl: poolRegistrationParams.metadata.metadataUrl,
```

Comment:
isn't this identity transformation?

Reply:
You are right, I am not sure why I did it that way, fixed thnx.